### ISSUE TITLE
Callback for logging forward, backward and update time

### ISSUE THREAD

**Issue by joshim5** [Lightning-AI/pytorch-lightning#19939]
### Description & Motivation

Cross-posting an interesting and important feature request from the discussions channel: https://github.com/Lightning-AI/pytorch-lightning/discussions/19928

Citing the discussion thread:
> I'm trying to track the performance of forward/backward/update time with a Callback. My current implementation is showing strange behavior. It seems that the callback order is (at least functionally) different when using gradient accumulation !=1. This is expected, but it's unclear how to handle both cases with a single callback.

### Pitch

Citing the discussion thread:
> I'd really appreciate help coming up with an almost-correct implementation for tracking (1) forward pass time, (2) backward pass time, (3) total time for an update (forward + backward + optimizer step) which might depend on gradient accumulation and (4) amount of time spent waiting on the dataloader to generate the next batch. Alternatively, for (3) I'm happy to track only optimizer.step time since this should tell me how long it's taking for devices to sync and gradients to update. I'm open to tracking related metrics or other metrics entirely as long as they're correlated with model throughput/performance.

### Alternatives

An implementation is proposed on the discussion thread, but the OP shows that it is in fact incorrect.

### Additional context

Please check out the discussion! https://github.com/Lightning-AI/pytorch-lightning/discussions/19928
